scvi/dataset/loom.py

```diff
@@ -1,86 +1,110 @@
+import logging
+import os
+
 import loompy
 import numpy as np
-import os
-import logging
-from .dataset import GeneExpressionDataset
 
+from scvi.dataset.dataset import DownloadableDataset
 
-class LoomDataset(GeneExpressionDataset):
-    r""" Loads a `.loom` file.
+logger = logging.getLogger(__name__)
 
-    Args:
-        :filename: Name of the `.loom` file.
-        :save_path: Save path of the dataset. Default: ``'data/'``.
-        :url: Url of the remote dataset. Default: ``None``.
+
+class LoomDataset(DownloadableDataset):
+    """Loads a `.loom` file.
 
     Examples:
         >>> # Loading a remote dataset
         >>> remote_loom_dataset = LoomDataset("osmFISH_SScortex_mouse_all_cell.loom", save_path='data/',
         ... url='http://linnarssonlab.org/osmFISH/osmFISH_SScortex_mouse_all_cells.loom')
         >>> # Loading a local dataset
         >>> local_loom_dataset = LoomDataset("osmFISH_SScortex_mouse_all_cell.loom", save_path='data/')
-
     """
 
-    def __init__(self, filename, save_path='data/', url=None):
-        self.download_name = filename
-        self.save_path = save_path
-        self.url = url
-
-        self.has_gene, self.has_batch, self.has_cluster = False, False, False
-
-        data, batch_indices, labels, gene_names, cell_types = self.download_and_preprocess()
-
-        X, local_means, local_vars, batch_indices_, labels = \
-            GeneExpressionDataset.get_attributes_from_matrix(data, labels=labels)
-        batch_indices = batch_indices if batch_indices is not None else batch_indices_
-        super().__init__(X, local_means, local_vars, batch_indices, labels,
-                         gene_names=gene_names, cell_types=cell_types)
-
-    def preprocess(self):
-        logging.info("Preprocessing dataset")
+    def __init__(
+        self,
+        filename: str,
+        save_path: str = "data/",
+        url: str = None,
+        delayed_populating: bool = False,
+    ):
+        super().__init__(
+            urls=url,
+            filenames=filename,
+            save_path=save_path,
+            delayed_populating=delayed_populating,
+        )
+
+    def populate(self):
+        logger.info("Preprocessing dataset")
         gene_names, labels, batch_indices, cell_types = None, None, None, None
-        ds = loompy.connect(os.path.join(self.save_path, self.download_name))
-        select = ds[:, :].sum(axis=0) > 0  # Take out cells that doesn't express any gene
-
-        if 'Gene' in ds.ra:
-            gene_names = ds.ra['Gene']
-
-        if 'BatchID' in ds.ca:
-            batch_indices = ds.ca['BatchID']
-            batch_indices = np.reshape(batch_indices, (batch_indices.shape[0], 1))[select]
-
-        if 'ClusterID' in ds.ca:
-            labels = np.array(ds.ca['ClusterID'])
+        ds = loompy.connect(os.path.join(self.save_path, self.filenames[0]))
+        select = (
+            ds[:, :].sum(axis=0) > 0
```

Comment:
Should this be here if we have a filter cells by count function?

Reply:
we can use the api

Reply:
I don't like it either but it was the original implementation. @romain-lopez do you think it is worth it to remove it and so changing the returned dataset? and so let the user do the filtering?

Reply:
@ANazaret we can load and use the API call is what I meant

Reply:
but it's more efficient this way so no reason to change it. @ANazaret we should keep this behaviour, to respect the contract that our database should be trainable directly, and for consistency with the rest of the codebase

Reply:
yes, I think for the cells we need to remove them (plus log a warning somewhere that we did that and which cells, especially if this is a dataset manually imported) otherwise we won't be able to train the model directly. And as @gabmis said, we want the models to be trainable directly.